src/libxtp/gwbse/gw.cc

```diff
@@ -300,5 +303,56 @@ void GW::CalculateHQP() {
   _Sigma_c.diagonal() = diag_backup;
 }
 
+void GW::PlotSigma(const Eigen::VectorXd& frequencies) const {
+  const Index steps = _opt.sigma_plot_steps;
+  const double spacing = _opt.sigma_plot_spacing;
+  const Index qptotal = _opt.qpmax - _opt.qpmin + 1;
+  XTP_LOG(Log::info, _log) << TimeStamp() << " Plotting Sigma diagonals "
+                           << std::flush;
+
+  std::vector<Index> ind;
+  if (_opt.sigma_plot_states == "all") {
+    ind.resize(qptotal);
+    std::iota(ind.begin(), ind.end(), _opt.qpmin);
+  } else {
+    tools::RangeParser rp;
+    rp.Parse(_opt.sigma_plot_states);
+    for (Index gw_level : rp) {
+      if (gw_level >= _opt.qpmin && gw_level <= _opt.qpmax) {
+        ind.push_back(gw_level);
+      }
+    }
+  }
+  const Index count = ind.size();
+
+  std::ofstream out;
+  out.open(_opt.sigma_plot_filename);
+
+  for (Index i = 0; i < count; i++) {
+    Index gw_level = ind[i];
+    out << boost::format("%1$somega(%2$d)\tsigma(%2$d)") %
+               (i == 0 ? "" : "\t") % gw_level;
+  }
+  out << std::endl;
+
+  boost::format numFormat("%+1.4f");
+  Eigen::IOFormat matFormat(Eigen::StreamPrecision, 0, "", "\t");
+  for (Index grid_point = 0; grid_point < steps; grid_point++) {
+    const double offset = (grid_point - ((steps - 1) / 2)) * spacing;
+    Eigen::VectorXd row = Eigen::VectorXd::Zero(2 * count);
+#pragma omp parallel for schedule(dynamic)
```

Comment:
can you parallelize over gridpoints not levels?

Reply:
Yeah, I could. But then I'd have to either (1) transpose the output matrix or (2) keep the entire output matrix in memory and only write when it's completely filled. Right now it writes the file row by row and only ever needs one row in memory. Though I suppose having a 10,000 x qptotal matrix in memory isn't that big of a deal. What do you think?

Reply:
actually to be honest I would scratch the "all" option and keep everything in memory. It is very unlikely that someone wants to print out so many states.